main.go

```diff
@@ -150,27 +195,209 @@ func readConfig(filename string) (*config, error) {
 	return &c, nil
 }
 
+// myClain define token struct
 type myClaim struct {
 	jwt.Claims
 	Roles []string `json:"roles,omitempty"`
 }
 
-func validateClaims(t string, sufficientRoles []string) error {
+// validateClaims check role fom config sufficetRoles is in token roles claim
+func validateClaims(t string, sufficientRoles []string, username string, addGroup bool, addmembership bool) error {
 	token, err := jwt.ParseSigned(t)
 	if err != nil {
 		return fmt.Errorf("error parsing token: %w", err)
 	}
 
 	claims := myClaim{}
 	if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
-		return fmt.Errorf("unable to extract claims from token:%w", err)
+		return fmt.Errorf("unable to extract claims from token: %w", err)
 	}
 	for _, role := range claims.Roles {
+
+		if addGroup {
+			_, err := createGroup(role, username)
+
+			if err != nil {
+				log.Printf("group: %s created" + role)
+			}
+		}
+
+		if addmembership {
+			_, err := addUserToGroup(role, username)
+
+			if err != nil {
+				log.Printf("membership: user %s to %s created", role, username)
+			}
+		}
+
 		for _, sr := range sufficientRoles {
 			if role == sr {
+				log.Print("validateClaims access granted role " + role + " is in token")
 				return nil
 			}
 		}
 	}
-	return fmt.Errorf("role:%s not found", sufficientRoles)
+	return fmt.Errorf("role: %s not found", sufficientRoles)
+}
+
+// createUser this create user is not exsits
+func createUser(username string) {
+	currentuser, err := unixuser.LookupUser(username)
+
+	if err != nil {
+		// if no user then add one
+		if currentuser == nil {
+
+			cmd := exec.Command("usr/sbin/useradd", "-m", "-s", "/bin/bash", "-c", app, username)
+			stdoutStderr, err := cmd.CombinedOutput()
+			if err != nil {
+				log.Printf("user already exists: %s ", err.Error())
```

Comment:
if useradd returns with an error, it can not be for sure that the user already exists. Need a helper func to detect if user is already present, which was done before already. --> Error message must be wrong, better return fmt.Errorf("unable to add user: %w", err)